### PR TITLE
Fix Non-dismissible notices #2100

### PR DESCRIPTION
**Pull request by raftaar1191**
## Description
PR for issues #2100 

## How Has This Been Tested?
Manually Tested

## Screenshots (jpeg or gifs if applicable):
image
image
image

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Related PR
https://github.com/WordImpress/Give-Gift-Aid/pull/66

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

File: includes/class-notices.php [hunk 1]
```diff
@@ -167,7 +167,7 @@ public function render_admin_notices() {
 
 			$css_id = ( false === strpos( $notice['id'], 'give' ) ? "give-{$notice['id']}" : $notice['id'] );
 
-			$css_class = "give-notice notice is-dismissible {$notice['type']} notice-{$notice['type']}";
+			$css_class = "give-notice notice ". ( 'error' === (string) $notice['type'] ? 'non': 'is' ) ."-dismissible {$notice['type']} notice-{$notice['type']}";
```
Comment (DevinWalker) [inline includes/class-notices.php]: Hmm, I don't think we want all `error` types as non-dismissible. We should discuss with @ravinderk

Comment (ravinderk) [reply includes/class-notices.php]: @DevinWalker @raftaar1191 I think we can create `is_dismissible` as notice params where by default this is set to `true` and developer can use this.

@DevinWalker Let me know what do you think?

**Comment by DevinWalker** [comment]
@raftaar1191 @ravinderk here's the Slack convo we had: https://givewp.slack.com/archives/C0FAGC83C/p1506051498000007

Transcript:

Devin Walker | Head of Product [8 minutes ago] 
@ravinder there's already `auto_dismissible` right?


Devin Walker | Head of Product [8 minutes ago] 
having both `is_dismissible` and `auto_dismissible` is confusing for developers


Devin Walker | Head of Product [7 minutes ago] 
there should just be `dismissible` with params `auto`, `true`, `false`


Devin Walker | Head of Product [7 minutes ago] 
default to `true`


Ravinder Kumar | Developer 
[7 minutes ago] 
That what I am thinking


Ravinder Kumar | Developer [7 minutes ago] 
So auto contain close icon also


Ravinder Kumar | Developer 
[6 minutes ago] 
Right?

Devin Walker | Head of Product [6 minutes ago] 
no because it goes away anyways


Devin Walker | Head of Product 
[6 minutes ago] 
no need to click it, unless for some reason someone has JS turned off


Devin Walker | Head of Product [6 minutes ago] 
but that's unlikely


Ravinder Kumar | Developer 
[6 minutes ago] 
Got it


Ravinder Kumar | Developer [6 minutes ago] 
Can you add this on pr


Devin Walker | Head of Product 
[5 minutes ago] 
Actually, auto dismiss probably should have an "x" icon by default


Ravinder Kumar | Developer [4 minutes ago] 
:grinning:


Devin Walker | Head of Product 
[4 minutes ago] 
We can add a new option here: https://github.com/WordImpress/Give/blob/de2da6c0378f967dac4f2246b0dd69b44c0edcb9/includes/class-notices.php#L501
GitHub
WordImpress/Give
Give - Democratizing Generosity - The best way to accept donations on WordPress
 


Devin Walker | Head of Product [3 minutes ago] 
`close_icon` set to `true` or `false` (edited)


Devin Walker | Head of Product 
[3 minutes ago] 
and also in backend notice


Ravinder Kumar | Developer [3 minutes ago] 
Checking


Ravinder Kumar | Developer 
[2 minutes ago] 
Yes, is_dismissble also doing same thinh

Ravinder Kumar | Developer [1 minute ago] 
Because as developer I know what it does in wordpress

**Comment by DevinWalker** [comment]
## Summary of Convo Above

@ravinderk @raftaar1191 let me know if you have any questions or comments on the items below. 

### Customize Give_Notices API

   1. Deprecate `auto_dismissible` option with backwards compatibility
    2. Create new `dismissible` option to replace it with the following options:
        1. `auto` - notices are removed automatically like with `auto_dismissible` set to true
        2. `false` no “x” icon at all, non-dismissible
        3. `true` dismissible icon visible and removes notices via AJAX and notice does not reappear on reload
    3. Customize https://github.com/WordImpress/Give/blob/de2da6c0378f967dac4f2246b0dd69b44c0edcb9/includes/class-notices.php#L501 and dismissible args with new option to conditionally output a dismissible icon for both frontend and backend notices
    4. Ensure that frontend errors can also be dismissed and closed with an icon as well. Confirm args work for both frontend and backend notices

**Comment by raftaar1191** [comment]
@DevinWalker @ravinderk 

Working on this as discuss with @ravinderk and suggested by @DevinWalker

**Comment by raftaar1191** [comment]
`dismissible` is set to `true` by deflaut

**Comment by raftaar1191** [comment]
Related PR: https://github.com/WordImpress/Give-Gift-Aid/pull/66

File: includes/class-notices.php [hunk 2]
```diff
@@ -99,6 +105,20 @@ public function register_notice( $notice_args ) {
...
+		 * Added to give support to backward compatibility to auto_dismissible.
+		 * Check if auto_dismissible is set and it true then unset and change dismissible parameter value to auto
+		 *
+		 * @since 1.8.14
+		 */
+		if ( isset( $notice_args['auto_dismissible'] ) ) {
+			if ( ! empty( $notice_args['auto_dismissible'] ) ) {
+				$notice_args['dismissible'] = 'auto';
+			}
+			// unset auto_dismissible as it has being deprecated.
```
Comment (DevinWalker) [inline includes/class-notices.php]: has `been`

File: includes/class-notices.php [hunk 3]
```diff
@@ -498,22 +537,41 @@ static function print_frontend_notice( $message, $echo = true, $notice_type = 'w
...
```
Comment (DevinWalker) [inline includes/class-notices.php]: Same typo

File: includes/class-notices.php [hunk 4]
```diff
@@ -498,22 +537,41 @@ static function print_frontend_notice( $message, $echo = true, $notice_type = 'w
...
 
 		$notice_args = wp_parse_args( $notice_args, $default_notice_args );
 
+		/**
+		 * Added to give support to backward compatibility to auto_dismissible.
+		 * Check if auto_dismissible is set and it true then unset and change dismissible parameter value to auto
+		 *
+		 * @since 1.8.14
+		 */
+		if ( isset( $notice_args['auto_dismissible'] ) ) {
```
Comment (ravinderk) [inline includes/class-notices.php]: @raftaar1191 I can see this code repeat multiple times. We can create a function which we handle deprecated functions.

Comment (raftaar1191) [reply includes/class-notices.php]: Yes we can do that

**Comment by ravinderk** [comment]
@raftaar1191 Nice PR

**Comment by raftaar1191** [comment]
@ravinderk  Thanks

**Pull request merged**